docs/docs/vm-fundamentals/builtins.md

```diff
@@ -4,4 +4,143 @@ sidebar_position: 2
 
 # Builtins
 
-Explain each builtin here. How do they work and what do they accomplish.
+Regular Cairo instructions are like regular Legos, they work, but for complex things they can be slow. Builtins are special Lego pieces that snap together easily, like pre-built wheels or doors. This makes building things much faster. In a Cairo context, builtins help write clean and efficient Cairo programs, just like these special Lego pieces help build cool things faster.
+
+This section provides an overview of the builtins functions available in the Cairo Virtual Machine (Cairo VM).
+
+### What is builtins in cairo?
+
+Builtins are predefined optimized low-level execution units which are added to the Cairo CPU board to perform predefined computations which are expensive to perform in Cairo.
+
+### Builtins and Cairo memory
+
+Communication between the CPU and built-in functionalities occurs via memory-mapped I/O. Each builtin is allocated a contiguous memory region and enforces specific constraints on the data residing within that area. The Pedersen builtin is a good example to explain this.
+
+Pedersen builtin establishes that:
+
+```
+[p + 2] = hash([p + 0], [p + 1])
+[p + 5] = hash([p + 3], [p + 4])
+[p + 8] = hash([p + 6], [p + 7])
+```
+
+The Cairo code may read or write from this memory cells to “invoke” the builtin. The following code verifies that `hash(x, y) == z`
+
+```
+// Write the value of x to [p + 0].
+x=[p]; // Write the value of y to [p + 1].
+y=[p + 1]; // The builtin makes that [p + 2] == hash([p + 0], [p + 1]).
+z=[p + 2];
+```
+
+Cairo memory immutability requires a careful usage of builtin instances because memory locations like `[p + 0]`, `[p + 1]`, and `[p + 2]` are used for a single hash computation, they cannot be reused for subsequent calculations and needs to tracking an unused memory location pointer (`hash_ptr`).
+
+By convention, the functions that utilizing builtins receive a pointer to the unused memory location as an argument and return an updated pointer reflecting the next available memory slot. For example:
+
+```
+func hash2(hash_ptr: felt*, x, y) -> (hash_ptr: felt*, z: felt) {
+  // Invoke the hash function.
+  x = [hash_ptr];
+  y = [hash_ptr + 1];
+  // Update pointer (increment by 3) and return result.
+  return (hash_ptr=hash_ptr + 3, z=[hash_ptr + 2]);
+}
+```
+
+On the other hand, Starkware Cairo library provides typed references for improved readability. Here's the code rewritten with `HashBuiltin` type like this example:
+
+```
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+
+func hash2(hash_ptr: HashBuiltin*, x, y) -> (hash_ptr: HashBuiltin*, z: felt) {
+  let hash = hash_ptr;
+  // Invoke the hash function using typed access.
+  hash.x = x;
+  hash.y = y;
+  // Update pointer and return result.
+  return (hash_ptr=hash_ptr + HashBuiltin.SIZE, z=hash.result);
+}
+```
+
+### Some specific builtins in Cairo VM
```

Comment:
The code in this section is still wrong and doesn't reflect the actual usages of these builtins. It would be good for readers to easily copy the code into a cairo file and execute it.

Reply:
@rodrigo-pino can you tell me what is exactly wrong in this code please? I was guided by a very similar example in the Cairo documentation and I can't find what is wrong

Reply:
Is not cairo code (except for output). I doubt you found this code in any documentation. Hope you can realize the obvious difference between the code the `output` builtin and the others

Reply:
Yep, I am not talking about the code above this comment, I am referring to the code examples of the subsection **Some specific builtins in Cairo VM**. As you'll notice, is very dissimilar to the code example above this comment which is the one I am not referring to. 

The only builtin of the builtins listed in this section which is an actual Cairo example is the `output` builtins. The remaining builtins, the ones defined below the `output` builtins, listed in order of appearance:
1. Pedersen
2. Range Check
3. ecdsa
4. bitwise

Are not valid examples of Cairo Code

Reply:
Now I understand it much better, thank you